### ISSUE TITLE
Federated extension script imrovements

### ISSUE THREAD

**Issue by martinRenou** [jupyterlab/jupyterlab#9421]
## Description

The federated extension script works well when the extension was generated with the cookiecutter, but it's not reliable in the case of a completely custom extension.

For example when running `jupyter labextension develop .`.

### 1. When the distribution name is not the same as the Python module name

The current script looks for `setup.py --name` to find the module name, that does not work if:
- there is a dash in the name (this is actually an issue in `traitlets.import_item`, we might want to replace it with `importlib.import_module` from now on)
- the distribution name is not the same as the module name

We should maybe look for the `packages` list found in the setup.py file, and maybe go over those all the packages to find the `_jupyter_labextension_paths` function, instead of relying on the `setup.py` name.

### 2. When the Python is imported from site-packages

In the case where the script successfully imported the module and found the `_jupyter_labextension_paths` function, it might happen that this module was actually imported from the installation path (`PREFIX/lib/python**/site-packages/module_name`).

In this case:
- the script tries to find the `package.json` file relatively to the Python module it found: https://github.com/jupyterlab/jupyterlab/blob/master/jupyterlab/federated_labextensions.py#L199
- it tries to find the labextension (`_jupyter_labextension_paths()['src']`) relatively to this Python module as well

This is problematic because those files are very likely to not be under `PREFIX/lib/python**/site-packages/module_name`, they are more likely to be in the development folder.

## Reproduce

You can reproduce those issues I mention in jupyterlab-lsp for example, in the PR I made for JupyterLab 3: https://github.com/krassowski/jupyterlab-lsp/pull/405

## Improvements

We might be able to improve this script by simplifying it. I was wondering if we could make something like:

`jupyter labextension develop --module module_name --package ./path/to/package.json`

This way the script knows where to found everything, the `module` parameter will allow to retrieve `_jupyter_labextension_paths` and get the `dest` from there (for the symlink in the case of `develop`), and the `package` parameter would contain everything that JupyterLab needs to know in order to build the package.

I might have a candide point of view, maybe it's more complicated than this. Please tell me if you see issues with this approach, I would be happy to open a PR if we agree on a better solution.

**Comment by jtpio** [comment]
Yes, this seems to be an issue in the common case where the distribution name contains `-`, while the module name is with `_`.

For example:

- distribution: `my-python-package`
- module: `my_python_package`

**Comment by jtpio** [comment]
As pointed out by Martin, it looks like this is also an issue when creating a new extension from the cookiecutter with a server extension (opt-in).

**Comment by martinRenou** [comment]
> As pointed out by Martin, it looks like this is also an issue when creating a new extension from the cookiecutter with a server extension (opt-in).

This might have been introduced by this PR https://github.com/jupyterlab/jupyterlab/pull/9326, so, sorry for that.

> We might be able to improve this script by simplifying it. I was wondering if we could make something like:
jupyter labextension develop --module module_name --package ./path/to/package.json

I guess my point here is that the script is trying to do too many things while not knowing so much about where to find the files it needs, and helping it find everything by providing the files in the command line might be better than trying to guess.

**Comment by fcollonval** [comment]
@martinRenou Is there some problems still unsolved?

**Comment by jtpio** [comment]
Linking to the PR that switched to using the `hatch` backend, which updated some of the logic for handling federated extensions: https://github.com/jupyterlab/jupyterlab/pull/12606.

With the extension cookiecutter being updated soon: https://github.com/jupyterlab/extension-cookiecutter-ts/pull/222

**Comment by jtpio** [comment]
Of course if none of the issues mentioned above still apply, we could also just close this issue and open new ones when new issues arise.

**Comment by JasonWeill** [comment]
Closing in favor of new issues to be filed for specific issues related to modern extension development, without the cookiecutter.
